dmoj/judge.py

```diff
@@ -459,69 +491,132 @@ def _grade_cases(self) -> Generator[Tuple[IPC, tuple], None, None]:
+
+            return failed_tests, GradeManyResult.OK
 
-                    # If the submission was killed due to a user-initiated abort, any result is meaningless.
-                    if self._abort_requested:
-                        yield IPC.GRADING_ABORTED, ()
-                        return
+        def should_run_pretest(case: TopLevelCase, failed_so_far: Set[int]) -> bool:
+            if case.dependencies is None:
+                # Default: depends on nothing.
+                return True
 
-                    if result.result_flag & Result.WA:
-                        # If we failed a 0-point case, we will short-circuit every case after this.
-                        is_short_circuiting_enabled |= not case.points
+            if case.dependencies & failed_so_far:
+                return False
 
-                        # Short-circuit if we just failed a case in a batch, or if short-circuiting is currently enabled
-                        # for all test cases (either this was requested by the site, or we failed a 0-point case in the
-                        # past).
-                        is_short_circuiting |= batch_number is not None or is_short_circuiting_enabled
+            return True
 
-                # Legacy hack: we need to allow graders to read and write `proc_output` on the `Result` object, but the
-                # judge controller only cares about the trimmed output, and shouldn't waste memory buffering the full
-                # output. So, we trim it here so we don't run out of memory in the controller.
-                result.proc_output = result.output
-                yield IPC.RESULT, (batch_number, case_number, result)
+        failed_pretests: Set[int] = set()
 
-            if batch_number:
-                if not is_short_circuiting:
-                    passed_batches.add(batch_number)
+        def should_run_main_test(case: TopLevelCase, failed_so_far: Set[int]) -> bool:
+            if case.dependencies is None:
+                # Default: depends on all pretests, and nothing else.
+                return not failed_pretests
 
-                yield IPC.BATCH_END, (batch_number,)
-                is_short_circuiting &= is_short_circuiting_enabled
+            if case.dependencies & failed_pretests:
+                return False
 
-        yield IPC.GRADING_END, ()
+            if case.dependencies & failed_so_far:
+                return False
+
+            return True
+
+        yield IPC.GRADING_BEGIN, (self.grader.run_pretests_only,)
+        pretests, main_tests = self.grader.cases()
+        try:
+            pretest_result = GradeManyResult.OK
+            if pretests:
+                yield IPC.PRETEST_BEGIN, ()
+                failed_pretests, pretest_result = yield from grade_many(pretests, should_run_pretest)
+                yield IPC.PRETEST_END, ()
+            if not self.grader.run_pretests_only:
+                if pretest_result == GradeManyResult.SKIP_ALL:
```

Comment:
Can this ever happen without `len(failed_pretests) != 0`? Do we need this extra enum?